packages/flutter/lib/src/cupertino/colors.dart

```diff
@@ -80,3 +93,672 @@ class CupertinoColors {
+  }
+
+  bool get _isInterfaceElevationDependent {
+    return color != elevatedColor
+        || darkColor != darkElevatedColor
+        || highContrastColor != highContrastElevatedColor
+        || darkHighContrastColor != darkHighContrastElevatedColor;
+  }
+
+  /// Resolves this `CupertinoDynamicColor` using the provided [BuildContext].
+  ///
+  /// Calling this method will create a new `CupertinoDynamicColor` that is almost
+  /// identical to this `CupertinoDynamicColor`, except the effective color is
+  /// changed to adapt to the given [BuildContext].
+  ///
+  /// For example, if the given [BuildContext] indicates the widgets in the subtree
+  /// should be displayed in dark mode (the surrounding [CupertinoTheme]'s [CupertinoThemeData.brightness]
+  /// or [MediaQuery]'s [MediaQueryData.platformBrightness] is [PlatformBrightness.dark]),
+  /// with a high accessibility contrast (the surrounding [MediaQuery]'s [MediaQueryData.highContrast]
+  /// is `true`), and an elevated interface elevation (the surrounding [CupertinoUserInterfaceLevel]'s
+  /// `data` is [CupertinoUserInterfaceLevelData.elevated]), the resolved
+  /// `CupertinoDynamicColor` will be the same as this [CupertinoDynamicColor],
+  /// except its effective color will be the `darkHighContrastElevatedColor` variant
+  /// from the orignal `CupertinoDynamicColor`.
+  ///
+  /// Calling this function may create dependencies on the closest instance of some
+  /// [InheritedWidget]s that enclose the given [BuildContext]. E.g., if [darkColor]
+  /// is different from [color], this method will call [CupertinoTheme.of], and
+  /// then [MediaQuery.of] if brightness wasn't specified in the theme data retrived
+  /// from the previous [CupertinoTheme.of] call, in an effort to determine the
+  /// brightness value.
+  ///
+  /// If any of the required dependecies are missing from the given context, an exception
+  /// will be thrown unless [nullOk] is set to `true`.
+  CupertinoDynamicColor resolveFrom(BuildContext context, { bool nullOk = false }) {
+    int brightnessNumber = 0;
+    int highContrastNumber = 0;
+    int interfaceElevationNumber = 0;
+
+    // If this CupertinoDynamicColor cares about brightness.
+    if (_isPlatformBrightnessDependent) {
+      final Brightness brightness = CupertinoTheme.brightnessOf(context, nullOk: nullOk) ?? Brightness.light;
+      brightnessNumber = brightness.index;
+    }
+
+    // If this CupertinoDynamicColor cares about accessibility contrast.
+    if (_isHighContrastDependent) {
+      final bool isHighContrastEnabled = MediaQuery.of(context, nullOk: nullOk)?.highContrast
+        ?? false;
+
+        highContrastNumber = isHighContrastEnabled ? 1 : 0;
+    }
+
+    // If this CupertinoDynamicColor cares about user interface elevation.
+    if (_isInterfaceElevationDependent) {
+      final CupertinoUserInterfaceLevelData level =
+        CupertinoUserInterfaceLevel.of(context, nullOk: nullOk)
```

Comment:
This expression could go on the previous line to be consistent with the previous code block